internal/checker/checker.go

```diff
@@ -21174,14 +21176,64 @@ func (c *Checker) instantiateTypeWithAlias(t *Type, m *TypeMapper, alias *TypeAl
 		c.error(c.currentNode, diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite)
 		return c.errorType
 	}
+	index := c.findActiveMapper(m)
+	if index == -1 {
+		c.pushActiveMapper(m)
+	}
+	var b KeyBuilder
+	b.WriteType(t)
+	b.WriteAlias(alias)
+	key := b.String()
+	cache := c.activeTypeMappersCaches[core.IfElse(index != -1, index, len(c.activeTypeMappersCaches)-1)]
+	if cachedType, ok := cache[key]; ok {
+		return cachedType
+	}
 	c.TotalInstantiationCount++
 	c.instantiationCount++
 	c.instantiationDepth++
 	result := c.instantiateTypeWorker(t, m, alias)
+	if index == -1 {
+		c.popActiveMapper()
+	} else {
+		cache[key] = result
+	}
 	c.instantiationDepth--
 	return result
 }
 
+var activeMapperCache = sync.Pool{
+	New: func() any {
+		return make(map[string]*Type, 1)
+	},
+}
+
+func (c *Checker) pushActiveMapper(mapper *TypeMapper) {
+	c.activeMappers = append(c.activeMappers, mapper)
+
+	c.activeTypeMappersCaches = append(c.activeTypeMappersCaches, activeMapperCache.Get().(map[string]*Type))
+}
+
+func (c *Checker) popActiveMapper() {
+	c.activeMappers[len(c.activeMappers)-1] = nil
+	c.activeMappers = c.activeMappers[:len(c.activeMappers)-1]
+
+	cache := c.activeTypeMappersCaches[len(c.activeTypeMappersCaches)-1]
+	c.activeTypeMappersCaches[len(c.activeTypeMappersCaches)-1] = nil
+	clear(cache)
+	activeMapperCache.Put(cache)
+	c.activeTypeMappersCaches = c.activeTypeMappersCaches[:len(c.activeTypeMappersCaches)-1]
+}
+
+func (c *Checker) findActiveMapper(mapper *TypeMapper) int {
+	return core.FindLastIndex(c.activeMappers, func(m *TypeMapper) bool { return m == mapper })
+}
+
+func (c *Checker) clearActiveMapperCaches() {
+	for _, cache := range c.activeTypeMappersCaches {
+		clear(cache)
```

Comment:
Maybe similar question here

Reply:
Since that would mean call stacks which have pushed onto them will then have an undefined map. Well, they wouldn't, but they'd have a reference to a cache that would never be used again.